src/feedbinapi.cpp

```diff
@@ -0,0 +1,321 @@
+}
+
+rsspp::Feed FeedbinApi::fetch_feed(const std::string& id)
+{
+	CurlHandle handle;
+	return fetch_feed(id, handle);
+}
+
+rsspp::Feed FeedbinApi::fetch_feed(const std::string& id,
+	CurlHandle& cached_handle)
+{
+	rsspp::Feed feed;
+	feed.rss_version = rsspp::Feed::FEEDBIN_JSON;
+
+	const json unread_entry_ids = run_op(FEEDBIN_UNREAD_ENTRIES_PATH, json());
+	if (!unread_entry_ids.is_array()) {
+		LOG(Level::ERROR,
+			"FeedbinApi::fetch_feed: unread_entry_ids is not an array");
+		return feed;
+	}
+
+	std::map<int, bool> unread_entry_id_map;
+	for (int entry_id : unread_entry_ids) {
+		unread_entry_id_map[entry_id] = true;
+	}
+
+	const std::string query = strprintf::fmt("/v2/feeds/%s/entries.json", id);
+
+	const json entries = run_op(query, json(), cached_handle, HTTPMethod::GET);
+	if (!entries.is_array()) {
+		LOG(Level::ERROR, "FeedbinApi::fetch_feed: entries is not an array");
+		return feed;
+	}
+
+	LOG(Level::INFO, "FeedbinApi::fetch_feed: %" PRIu64 " items",
+		static_cast<uint64_t>(entries.size()));
+	try {
+		for (const auto& entry : entries) {
+			rsspp::Item item;
+
+			if (!entry["title"].is_null()) {
+				item.title = entry["title"];
+			}
+
+			if (!entry["url"].is_null()) {
+				item.link = entry["url"];
+			}
+
+			if (!entry["author"].is_null()) {
+				item.author = entry["author"];
+			}
+
+			if (!entry["content"].is_null()) {
+				item.content_encoded = entry["content"];
+			}
+
+			const int64_t entry_id = entry["id"];
+			item.guid = std::to_string(entry_id);
+
+			item.pubDate = entry["published"];
+
+			if (unread_entry_id_map[entry_id]) {
+				item.labels.push_back("feedbin:unread");
+			} else {
+				item.labels.push_back("feedbin:read");
+			}
+
+			feed.items.push_back(item);
+		}
+	} catch (json::exception& e) {
+		LOG(Level::ERROR, "Exception occurred while parsing feeed: ", e.what());
+	}
+
+	std::sort(feed.items.begin(), feed.items.end(),
+	[](const rsspp::Item &a, const rsspp::Item &b) {
+		return a.pubDate_ts > b.pubDate_ts;
+	});
+
+	return feed;
+}
+
+bool FeedbinApi::update_article_flags(const std::string& oldflags,
+	const std::string& newflags,
+	const std::string& guid)
+{
+	std::string star_flag = cfg.get_configvalue("feedbin-flag-star");
+	bool success = true;
+
+	if (star_flag.length() > 0) {
+		update_flag(oldflags, newflags, star_flag[0],
+		[&](bool added) {
+			success = star_article(guid, added);
+		});
+	}
+
+	return success;
+}
+
+void FeedbinApi::add_custom_headers(curl_slist** /* custom_headers */)
+{
+	// nothing required
+}
+
+bool FeedbinApi::star_article(const std::string& article_id, bool star)
+{
+	std::vector<std::string> ids;
+	ids.push_back(article_id);
+	json body;
+	body["starred_entries"] = ids;
+
+	CurlHandle handle;
+
+	HTTPMethod method = star ? HTTPMethod::POST : HTTPMethod::DELETE;
+
+	curl_slist* headers = NULL;
+	headers = curl_slist_append(headers, "Content-Type: application/json");
+	curl_easy_setopt(handle.ptr(), CURLOPT_HTTPHEADER, headers);
```

Comment:
Apparently this is done for every POST and DELETE, so perhaps this code could be moved to `run_op`?